### ISSUE TITLE
Register for binlog events

### ISSUE THREAD

**Issue by normanrz** [blackbeam/mysql_async#99]
In my project, I need a listener for MySQL's binlog events. There are [2 crates available](https://crates.io/search?q=binlog) that can parse the Binlog event data, but there is no network-based listener. I think this could be a nice addition to this crate.

Over the weekend, I built a prototype that adds commands for registering to the binlog events. I am happy to work on this further in a PR, but first wanted to get your feedback on this. See: https://github.com/blackbeam/mysql_async/compare/master...normanrz:expose-binlog

The API is as follows: The client code creates a `Conn`, registers that conn as a slave and then requests binlog events. After that, the client can repeatedly call `read_packet` to receive the next binlog event, which needs to be parsed and used. The conn can't be used for regular queries anymore. 

Here is a runnable test: https://github.com/normanrz/mysql_async/blob/expose-binlog/src/binlog/mod.rs
Please note that mysqld needs to be configured for binlogs: https://github.com/normanrz/mysql_async/blob/expose-binlog/.circleci/config.yml#L17 
In my example code, I use the `mysql_binlog` crate for parsing. There is [a PR for my changes](https://github.com/EasyPost/rust-mysql-binlog/pull/1) in that crate.

Would you rather implement the command code for `COM_REGISTER_SLAVE` and `COM_BINLOG_DUMP` in `mysql_common`?
Also, I couldn't figure out how to expose `read_packet` for `Conn` as a public function. 

My implementation is built upon @dannysu's fork: https://github.com/blackbeam/rust-mysql-simple/compare/master...dannysu:binlogsync

**Comment by blackbeam** [comment]
Hi. Sorry for the delayed response.

> Would you rather implement the command code for `COM_REGISTER_SLAVE` and `COM_BINLOG_DUMP` in mysql_common?

I'm currently working on it in the `binlog` branch of rust_mysql_common (see [here][1]).

> Also, I couldn't figure out how to expose read_packet for Conn as a public function.

I think I'll create a separate structure for this.

As a next step, I'll try to implement your runnable test on top of it.

[1]: https://github.com/blackbeam/rust_mysql_common/pull/18

**Comment by normanrz** [comment]
Cool, I'm happy to see you implementing the binlog functionality in rust_mysql_common. Binlog listening support is the final missing piece for me to move my service from node to rust.
Let me know if there is anything I can help with.

**Comment by blackbeam** [comment]
Hi.

[This branch][1] of `mysql-async` implements `Conn::get_binlog_stream` as a way to listen for binlog events (please see [this test][2]). `Conn::get_binlog_stream` takes the [`BinlogRequest`][3] struct as an argument, which is similar to `MYSQL_RPL` struct of a MySql server.

Could you please see if it fits your needs?

[1]: https://github.com/blackbeam/mysql_async/tree/binlog-stream
[2]: https://github.com/blackbeam/mysql_async/blob/binlog-stream/src/conn/mod.rs#L1011
[3]: https://github.com/blackbeam/mysql_async/blob/binlog-stream/src/conn/mod.rs#L851

**Comment by normanrz** [comment]
It looks really exciting! It is a very nice API. Thank you so much for your work on this.
I‘ll try to integrate your branch in my app this weekend and will get back with my comments.

**Comment by normanrz** [comment]
One thing I noticed, the return type of `conn.get_binlog_stream` is quite bulky (`Result<Pin<Box<dyn futures_core::stream::Stream<Item = Result<mysql_common::binlog::Event>>>>>`). Would it be possible to export a type alias for that?

**Comment by normanrz** [comment]
2 more issues:
- Recent MariaDB versions still send the [V1 row events](https://mariadb.com/kb/en/rows_event_v1/). Would be great if they were parsed as well. 
- I tried reading the stream from a spawned tokio task, but couldn't get it to work. I prepared a small code sample for that: https://github.com/normanrz/test-binlog-rs

**Comment by normano** [comment]
The binlog parsing looks good from what I am seeing so far. Using MySQL 8.

**Comment by normanrz** [comment]
I have to say that the new mutable connection API very nice to work with! The binlog request is also straightforward.

Working with the binlog events still feels very low-level. I think it would make sense to have a high-level API on top of the low-level API. I believe the most frequent use case is to process the write/update/delete events with the respective values. It would be great if these events were parsed in a way that the individual values were rust types. Maybe I am overlooking something here?

**Comment by normano** [comment]
I think was the ticket author said should be in a different ticket (if the library owner wants this also) or new different library (presumably not created by the author of this library).

To parse row values requires knowing table information outside of TableMapEvents. How do you distinguish between VARCHAR and VARBINARY when MySQL treats them the same internally as a char[] (string of uninterpreted bytes)? I looked at Easypost's Mysql binlog file parser and the implementation is unfinished and makes assumptions that are not consistent with MySQL. The ticket author has an open PR with Easypost's libary but no traction, so I will not contribute what I've done there.

**Comment by blackbeam** [comment]
@normanrz 

> One thing I noticed, the return type of conn.get_binlog_stream is quite bulky. Would it be possible to export a type alias for that?

I've introduced the `BinlogStream` structure in 974b052.

> I tried reading the stream from a spawned tokio task, but couldn't get it to work.

Works for me with slight modifications (notably an `await` on the result of `tokio::spawn`):

```diff
diff --git a/src/main.rs b/src/main.rs
index 7a49e14..2e00c37 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -1,6 +1,6 @@
 use futures_core;
 use futures_util::stream::StreamExt;
-use mysql_async::prelude::*;
+use mysql_async::{*, prelude::*};
 use mysql_common;
 use std::pin::Pin;

@@ -10,24 +10,16 @@ type Result<T> = std::result::Result<T, Error>;
 async fn get_binlog_stream(
     url: &str,
     server_id: u32,
-) -> Result<
-    Pin<
-        Box<
-            dyn futures_core::stream::Stream<
-                Item = mysql_async::Result<mysql_common::binlog::Event>,
-            >,
-        >,
-    >,
-> {
+) -> Result<BinlogStream> {
     let opts = mysql_async::Opts::from_url(url)?;
     let mut conn = mysql_async::Conn::new(opts).await?;
-    let (filename, position): (String, u64) = "SHOW BINARY LOGS".first(&mut conn).await?.unwrap();
+    let (filename, position, _): (String, u64, Value) = "SHOW BINARY LOGS".first(&mut conn).await?.unwrap();

     let binlog_stream = conn
         .get_binlog_stream(
             mysql_async::BinlogRequest::new(server_id)
                 .with_filename(filename)
-                .with_pos(position),
+                .with_pos(4_u64),
         )
         .await?;
     return Ok(binlog_stream);
@@ -35,26 +27,26 @@ async fn get_binlog_stream(

 #[tokio::main]
 async fn main() -> Result<()> {
-    let database_url = "mysql://root:password@localhost/mysql";
+    let database_url = "mysql://root:password@localhost:3307/mysql";

     tokio::spawn(async move {
         let mut i = 0;
         let mut binlog_stream = get_binlog_stream(database_url, 12).await.unwrap();
         while let Ok(event) = binlog_stream.next().await.unwrap() {
+            i += 1;
             let event_data = event.read_data().unwrap().unwrap();
             match event_data {
                 mysql_common::binlog::EventData::WriteRowsEventV1(event_data) => {
                     println!("{:?}", event_data);
-                    i += 1;
                 }
                 _ => {}
             }
-            // println!("{:?}", event.header.event_type.get().unwrap());
+            println!("{:?}", event.header.event_type.get().unwrap());
             if i > 100 {
                 break;
             }
         }
-    });
+    }).await?;

     Ok(())
 }
```

> Recent MariaDB versions still send the V1 row events. Would be great if they were parsed as well.

Ok. I'll add it.

**Comment by normanrz** [comment]
Thanks for helping me debug my tokio issue. Works like a charm now. Also, thanks for the `BinlogStream` struct.

> To parse row values requires knowing table information outside of TableMapEvents. How do you distinguish between VARCHAR and VARBINARY when MySQL treats them the same internally as a char[] (string of uninterpreted bytes)? I looked at Easypost's Mysql binlog file parser and the implementation is unfinished and makes assumptions that are not consistent with MySQL. The ticket author has an open PR with Easypost's libary but no traction, so I will not contribute what I've done there.

This depends on where @blackbeam wants to take his library. I still think that a major use case for working with the binlog is to read and understand row events. With this API it is now possible to get them (which is awesome!), but they are not ready to work with because still encoded in a blob. 

I would've imagined an API similar to `mysql_async::from_row` to parse the row event contents, where the library user specifies how to parse the contents. Of course it would also be possible to add this row parsing as a library on top.

**Comment by blackbeam** [comment]
Hi, @normanrz, @normano.

I've updated `mysql_async/binlog-stream` and `mysql_comon/binlog`:

* v1 rows are supported now ([see here][1]),
* now you can iterate over rows of a row event given the corresponding `TableMapEvent` ([see here][2])
* blackbeam/rust_mysql_common#22 is fixed

It's not yet finished, because:
1. there is no support of partial events,
2. there is no support of JSON columns.

so the API will change, but this change won't be dramatic.

[1]: https://github.com/blackbeam/mysql_async/blob/binlog-stream/src/conn/mod.rs#L848-L850
[2]: https://github.com/blackbeam/mysql_async/blob/binlog-stream/src/conn/mod.rs#L851-L853

**Comment by normano** [comment]
When you say partial events, you mean when optional metadata system option is not set to FULL since you would not be able to get column names without it?

**Comment by blackbeam** [comment]
@normano, I meant [PARTIAL_UPDATE_ROWS_EVENT][1].

[1]: https://github.com/mysql/mysql-server/blob/8.0/sql/rpl_record.cc#L471-L474

**Comment by normano** [comment]
Ok, I'll test this out as soon as possible

**Comment by normano** [comment]
I took a look at this and I saw the column names are not parsed during read row. Makes sense when system option binlog_row_metadata is not set to FULL.

On every iteration on RowsEventRows, it'll parse the columns. I saw that one could parse it in the table map event to do column processing work once. TableMap could still be a hashmap but with <u64, Table>. Table would be something like {db_name, table_name, columns, null_bitmask, optional_metadata, etc.}

**Comment by normanrz** [comment]
@blackbeam Your work is very impressive. I have all I need for my use case now. The addition of the row parsing is immensely helpful. As always, the code is clean and API is easy to understand when digging through the tests and code.

For beginners, some documentation would be helpful. I am happy to contribute if you wish.

As an example, my code looks something like this:
```rust
type PointRow = (String, f64, f64, f64, Option<f64>, Option<f64>);

enum RowEvent<T: FromRow> {
  NewRow { new_row: T },
  DeletedRow { old_row: T },
  UpdatedRow { old_row: T, new_row: T },
}

async fn connect_binlog_stream(
  database_url: &str,
  server_id: u32,
) -> Result<mysql_async::BinlogStream> {
  let opts = mysql_async::Opts::from_url(database_url)?;
  let mut conn = mysql_async::Conn::new(opts).await?;
  let (filename, position): (String, u64) = conn
    .query_iter("SHOW BINARY LOGS")
    .await?
    .reduce_and_drop(
      (String::from("mysqld-bin.000001"), 0),
      |_, (filename, position): (String, u64)| (filename, position),
    )
    .await?;

  let binlog_stream = conn
    .get_binlog_stream(
      mysql_async::BinlogRequest::new(server_id)
        .with_filename(filename)
        .with_pos(position),
    )
    .await?;
  Ok(binlog_stream)
}

async fn process() -> Result<()> {
  let database_url = "mysql://...";
  let server_id: u32 = 12;
  let database_name = "test";
  let table_name = "points";

  let point_cache = PointCache::new();

  let mut table_map = HashMap::new();
  let mut binlog_stream = connect_binlog_stream(database_url, server_id).await?;

  while let Some(Ok(event)) = binlog_stream.next().await {
    let event_data = event.read_data()?.unwrap();
    match event_data {
      mysql_common::binlog::EventData::WriteRowsEvent(mysql_common::binlog::WriteRowsEvent(re))
      | mysql_common::binlog::EventData::WriteRowsEventV1(
        mysql_common::binlog::WriteRowsEventV1(re),
      ) => {
        let table_map_item = table_map.get(&re.table_id).unwrap();
        if table_map_item.database_name.get() == database_name
          && table_map_item.table_name.get() == table_name
        {
          for row in re.rows(table_map_item) {
            let (_, new_row) = row.unwrap();
            let new_row_data: PointRow = mysql_async::from_row(new_row.unwrap());
            point_cache.apply(RowEvent::<PointRow>::NewRow {
              new_row: new_row_data,
            });
          }
        }
      }
      mysql_common::binlog::EventData::UpdateRowsEvent(mysql_common::binlog::UpdateRowsEvent(
        re,
      ))
      | mysql_common::binlog::EventData::UpdateRowsEventV1(
        mysql_common::binlog::UpdateRowsEventV1(re),
      ) => {
        let table_map_item = table_map.get(&re.table_id).unwrap();
        if table_map_item.database_name.get() == database_name
          && table_map_item.table_name.get() == table_name
        {
          for row in re.rows(table_map_item) {
            let (old_row, new_row) = row.unwrap();
            let old_row_data: PointRow = mysql_async::from_row(old_row.unwrap());
            let new_row_data: PointRow = mysql_async::from_row(new_row.unwrap());
            point_cache.apply(RowEvent::<PointRow>::UpdatedRow {
              old_row: old_row_data,
              new_row: new_row_data,
            });
          }
        }
      }
      mysql_common::binlog::EventData::DeleteRowsEvent(mysql_common::binlog::DeleteRowsEvent(
        re,
      ))
      | mysql_common::binlog::EventData::DeleteRowsEventV1(
        mysql_common::binlog::DeleteRowsEventV1(re),
      ) => {
        let table_map_item = table_map.get(&re.table_id).unwrap();
        if table_map_item.database_name.get() == database_name
          && table_map_item.table_name.get() == table_name
        {
          for row in re.rows(table_map_item) {
            let (old_row, _) = row.unwrap();
            let old_row_data: PointRow = mysql_async::from_row(old_row.unwrap());
            point_cache.apply(RowEvent::<PointRow>::DeletedRow {
              old_row: old_row_data,
            });
          }
        }
      }
      mysql_common::binlog::EventData::TableMapEvent(tme) => {
        table_map.insert(tme.table_id, tme);
      }
      _ => (),
    };
  }
  Ok(())
}
```

**Comment by normanrz** [comment]
@blackbeam Thanks again for building the binlog support! I am already using it productively in one of my projects. Are you planning on merging and releasing the functionality?

**Comment by blackbeam** [comment]
@normanrz, I'm glad to hear that. The implementation is still dirty and inefficient (mentioned by @normano). It also lacks some features (described earlier). I want to fix all this before merging and releasing, but to be honest, that won't be fast (I'm quite busy ATM).

Anyway, I'll try to keep this branch up to date with the latest master.

**Comment by normanrz** [comment]
Thanks for your answer and thanks for supporting this branch. I'm very grateful!

**Comment by blackbeam** [comment]
Published in [v0.28.0](https://github.com/blackbeam/mysql_async/releases/tag/v0.28.0)
